File: Robot_Control_Course/Chapter6/Lesson5/Chapter6_Lesson5.java

public class Interaction1DOF {

    public enum Mode { IMPEDANCE, ADMITTANCE }

    public static class State {
        public double x;
        public double v;
        public double xCommand;
        public double vVirtual;
    }

    // Parameters
    private final double Mr = 5.0;
    private final double Dr = 2.0;
    private final double Ke = 500.0;
    private final double De = 10.0;

    private final double Md = 2.0;
    private final double Dd = 30.0;
    private final double Kd = 200.0;

    public void step(State s, double xRef, double dt, Mode mode) {
        double Fenv = Ke * s.x + De * s.v;
        double Fu;

        if (mode == Mode.IMPEDANCE) {
            double e  = xRef - s.x;
            double ed = 0.0 - s.v;
            Fu = Kd * e + Dd * ed;

            double a = (Fu + Fenv - Dr * s.v) / Mr;
            s.v += dt * a;
            s.x += dt * s.v;

        } else {
            // Admittance
            double ac = (Fenv - Dd * s.vVirtual - Kd * (s.xCommand - xRef)) / Md;
            s.vVirtual += dt * ac;
            s.xCommand += dt * s.vVirtual;

            double ePos = s.xCommand - s.x;
            Fu = 1000.0 * ePos; // inner position loop

            double a = (Fu + Fenv - Dr * s.v) / Mr;
            s.v += dt * a;
            s.x += dt * s.v;
        }
    }
}

File: Robot_Control_Course/Chapter6/Lesson5/Chapter6_Lesson5.cpp

#include <iostream>
#include <cmath>

enum class Mode { Impedance, Admittance };

struct InteractionState {
    double x;    // position
    double v;    // velocity
    double x_c;  // commanded position (admittance)
};

void stepInteraction(InteractionState& s,
                     double x_d,
                     double dt,
                     Mode mode)
{
    // Robot and environment parameters
    const double M_r = 5.0;
    const double D_r = 2.0;
    const double K_e = 500.0;
    const double D_e = 10.0;

    const double M_d = 2.0;
    const double D_d = 30.0;
    const double K_d = 200.0;

    // Environment force
    double F_env = K_e * s.x + D_e * s.v;

    double F_u = 0.0;

    if (mode == Mode::Impedance) {
        double e  = x_d - s.x;
        double ed = 0.0 - s.v;

        F_u = K_d * e + D_d * ed;

        double a = (F_u + F_env - D_r * s.v) / M_r;
        s.v += dt * a;
        s.x += dt * s.v;

    } else { // Admittance
        static double v_c = 0.0;  // virtual velocity
        double a_c = (F_env - D_d * v_c - K_d * (s.x_c - x_d)) / M_d;
        v_c += dt * a_c;
        s.x_c += dt * v_c;

        // Inner position servo
        double e_pos = s.x_c - s.x;
        F_u = 1000.0 * e_pos;

        double a = (F_u + F_env - D_r * s.v) / M_r;
        s.v += dt * a;
        s.x += dt * s.v;
    }
}
